onecodex/api.py

```diff
@@ -1,31 +1,65 @@
 from __future__ import print_function
-from datetime import datetime
-import errno
 import filelock
 import json
 import logging
 import os
+import requests
 from requests.adapters import HTTPAdapter
 from requests.auth import HTTPBasicAuth
 from requests.packages.urllib3.util.retry import Retry
 import warnings
 
-from onecodex.exceptions import OneCodexException
 from onecodex.lib.auth import BearerTokenAuth
 from onecodex.utils import collapse_user, init_sentry
-from onecodex.vendored.potion_client import Client as PotionClient
-from onecodex.vendored.potion_client.converter import (
-    PotionJSONSchemaDecoder,
-    PotionJSONDecoder,
-    PotionJSONEncoder,
-)
-from onecodex.vendored.potion_client.utils import upper_camel_case
 from onecodex.version import __version__
 
 
 log = logging.getLogger("onecodex")
 
 
+class HTTPClient:
+    """Simple requests-based HTTP client for API requests."""
+
+    def __init__(self, auth=None, headers=None, timeout=30.0, **kwargs):
+        self.session = requests.Session()
+        if auth:
+            self.session.auth = auth
+        if headers:
+            self.session.headers.update(headers)
+
+        # Set backoff / retry strategy for 429s
+        retry_strategy = Retry(
+            total=3, backoff_factor=4, allowed_methods=None, status_forcelist=[429, 502, 503]
+        )
+        adapter = HTTPAdapter(max_retries=retry_strategy)
+        self.session.mount("http://", adapter)
+        self.session.mount("https://", adapter)
+
+    def get(self, url, **kwargs):
+        """Make GET request."""
+        return self.session.get(url, **kwargs)
+
+    def post(self, url, **kwargs):
+        """Make POST request."""
+        return self.session.post(url, **kwargs)
+
+    def patch(self, url, **kwargs):
+        """Make PATCH request."""
+        return self.session.patch(url, **kwargs)
+
+    def put(self, url, **kwargs):
+        """Make PUT request."""
+        return self.session.put(url, **kwargs)
+
+    def delete(self, url, **kwargs):
+        """Make DELETE request."""
+        return self.session.delete(url, **kwargs)
+
+    def request(self, method, url, **kwargs):
```

Comment:
TODO: Maybe make these all flow through `request` and then we can have error handling here vs. doing `resp.raise_for_status` everywhere

Reply:
So we don't _really_ have unified error handling, so perhaps this is a follow up TODO. I've made everything use `request` at least for simplicity / ease of doing that later. I have a commit that adds a special `raise_for_status=True` kwarg support, but honestly it feels a little gross. 🤷‍♂️

Reply:
Yeah I fixed that and reverted the `raise_for_status` kwarg. 🙇‍♂️